### PR TITLE
[WIP]multi-cluster: create/show warehouse; alter warehouse add/remove cluster; show clusters; alter warehouse properties

### PR DESCRIPTION
**Pull request by abc982627271**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/warehouse/Cluster.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+public class Cluster {
+    @SerializedName(value = "id")
+    private long id;
+    @SerializedName(value = "wgid")
+    private long workerGroupId;
+
+    // 注：Warehouse 和 Cluster 对象只记录了running 和 pending sql 的数量,
+    // 我们假设 sql 排队功能 和 Cluster 类完全解耦，Cluster 类提供接口,
+    // sql 排队功能根据 sql 执行情况负责调用这些接口更新 counter
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/warehouse/Cluster.java:29]: please in english

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocksLex.g4 [hunk 2]
```diff
@@ -357,6 +357,8 @@ VARCHAR: 'VARCHAR';
 VARIABLES: 'VARIABLES';
 VERBOSE: 'VERBOSE';
 VIEW: 'VIEW';
+WAREHOUSE: 'WAREHOUSE';
+WAREHOUSES: 'WAREHOUSES';
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocksLex.g4:361]: should put it in nonReserved

File: fe/fe-core/src/main/java/com/starrocks/warehouse/Warehouse.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    public void writeLock() {
+        long startMs = TimeUnit.MILLISECONDS.convert(System.nanoTime(), TimeUnit.NANOSECONDS);
+        Thread formerOwner = rwLock.getOwner();
+        this.rwLock.writeLock().lock();
+        logSlowLockEventIfNeeded(startMs, "writeLock", formerOwner);
+    }
+
+    public void writeUnlock() {
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/warehouse/Warehouse.java:114]: Is this lock necessary?
I think you can put the writelock in the warehouse manager

File: fe/fe-core/src/main/java/com/starrocks/connector/ConnectorMetadata.java [hunk 4]
```diff
@@ -56,6 +56,15 @@ default List<String> listDbNames() {
         return Lists.newArrayList();
     }
 
+    /**
+     * List all warehouse names of connector
+     *
+     * @return a list of string containing all database names of connector
+     */
+    default List<String> listWhNames() {
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/connector/ConnectorMetadata.java:64]: I don't think we need to implement `listWhNames()` and `createWarehouse()` in connector, computing warehouse has nothing to do with sr connector.

File: fe/fe-core/src/main/java/com/starrocks/qe/ShowExecutor.java [hunk 5]
```diff
@@ -518,6 +521,34 @@ private void handleShowProc() throws AnalysisException {
...
+        List<String> whNames = metadataMgr.listWhNames();
+
+        PatternMatcher matcher = null;
+        if (showStmt.getPattern() != null) {
+            matcher = PatternMatcher.createMysqlPattern(showStmt.getPattern(),
+                    CaseSensibility.DATABASE.getCaseSensibility());
+        }
+        Set<String> whNameSet = Sets.newTreeSet();
+        for (String whName : whNames) {
+            // Filter dbname
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/qe/ShowExecutor.java:537]: warehouse name ?

File: fe/fe-core/src/main/java/com/starrocks/server/MetadataMgr.java [hunk 6]
```diff
@@ -83,6 +90,19 @@ public void removeQueryMetadata() {
         }
     }
 
+    public List<String> listWhNames() {
+        Optional<ConnectorMetadata> connectorMetadata = getWarehouseMetaData();
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/server/MetadataMgr.java:94]: Warehouse has nothing to do with connector api, actually i think the best metadata object to refer to is `Catalog`, you can refer to `handleShowCatalogs()` to see how to implement `show warehouses`;

File: fe/fe-core/src/main/java/com/starrocks/qe/ShowExecutor.java [hunk 7]
```diff
@@ -518,6 +521,34 @@ private void handleShowProc() throws AnalysisException {
         resultSet = new ShowResultSet(metaData, finalRows);
     }
 
+    // show warehouse statement
+    private void handleShowWh() throws AnalysisException, DdlException {
+        ShowWhStmt showStmt = (ShowWhStmt) stmt;
+        List<List<String>> rows = Lists.newArrayList();
+        List<String> whNames = metadataMgr.listWhNames();
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/qe/ShowExecutor.java:528]: plz refactor this method to use api of `WarehouseMgr` class directly.

File: fe/fe-core/src/main/java/com/starrocks/server/WarehouseManager.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class WarehouseManager implements Writable, ConnectorMetadata {
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/server/WarehouseManager.java:39]: plz rename to `WarehouseMgr` to be consistent with the convention

File: fe/fe-core/src/main/java/com/starrocks/connector/ConnectorMetadata.java (hunk 4 above)
Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/connector/ConnectorMetadata.java:64]: we should remove this function

File: fe/fe-core/src/main/java/com/starrocks/connector/ConnectorMetadata.java [hunk 9]
```diff
@@ -121,6 +130,10 @@ default void clear() {
     default void refreshTable(String srDbName, Table table, List<String> partitionNames) {
     }
 
+    default void createWarehouse(String whName) throws DdlException, AlreadyExistsException {
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/connector/ConnectorMetadata.java:133]: we should remove this function

File: fe/fe-core/src/main/java/com/starrocks/qe/ShowExecutor.java [hunk 10]
```diff
@@ -1796,6 +1800,37 @@ private void handleShowCatalogs() {
         resultSet = new ShowResultSet(showCatalogsStmt.getMetaData(), rowSet);
     }
 
+    // show warehouse statement
+    private void handleShowWarehouses() throws AnalysisException, DdlException {
+        ShowWhStmt showStmt = (ShowWhStmt) stmt;
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/qe/ShowExecutor.java:1805]: I think we should add test case for `create warehouse` and `show warehouses`, add all the test case in the end may not be a good idea.

File: fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java [hunk 11]
```diff
@@ -1310,6 +1318,8 @@ public void loadImage(String imageDir) throws IOException, DdlException {
             remoteChecksum = dis.readLong();
             loadRBACPrivilege(dis);
             checksum = MVManager.getInstance().reload(dis, checksum);
+            checksum = warehouseMgr.loadWarehouses(dis, checksum);
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java:1321]: for compatibility reason, plz put this before `loadRBACPrivilege()`

File: fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java [hunk 12]
```diff
@@ -1576,6 +1586,8 @@ public void saveImage(File curFile, long replayedJournalId) throws IOException {
             dos.writeLong(checksum);
             saveRBACPrivilege(dos);
             checksum = MVManager.getInstance().store(dos, checksum);
+            checksum = warehouseMgr.saveWarehouses(dos, checksum);
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java:1589]: for compatibility reason, plz put this before saveRBACPrivilege()

File: fe/fe-core/src/main/java/com/starrocks/server/WarehouseManager.java [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class WarehouseManager implements Writable {
+    private static final Logger LOG = LogManager.getLogger(WarehouseManager.class);
+
+    @SerializedName(value = "idToWh")
+    private final ConcurrentHashMap<Long, Warehouse> idToWh = new ConcurrentHashMap<>();
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/server/WarehouseManager.java:52]: if we use rw lock, we don't need to use ConcurrentHashMap, normal hashMap is enough

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/WarehouseAnalyzer.java [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        @Override
+        public Void visitCreateWarehouseStatement(CreateWarehouseStmt statement, ConnectContext context) {
+            String whName = statement.getFullWhName();
+            if (Strings.isNullOrEmpty(whName)) {
+                throw new SemanticException("'warehouse name' can not be null or empty");
+            }
+            try {
+                FeNameFormat.checkCatalogName(whName);
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/WarehouseAnalyzer.java:43]: should be checkWarehouseName?

**Comment by nshangyiming** [comment]
plz rename this pr to a simple name and put the details in the "problem summary" section

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 15]
```diff
@@ -759,6 +797,20 @@ partitionRenameClause
     : RENAME PARTITION parName=identifier newParName=identifier
     ;
 
+// ---------Alter warehouse clause---------
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:800]: we don't need to define this?

File: fe/fe-core/src/main/java/com/starrocks/warehouse/Warehouse.java [hunk 16]
```diff
@@ -0,0 +1,338 @@
...
+            clusters.remove(cluster.getId());
+            // for debug
+            LOG.info("replay remove cluster {} for warehouse {} ", cluster.getId(), name);
+        } finally {
+            writeUnLock();
+        }
+    }
+
+    public void suspend() {
+        state = WarehouseState.SUSPENDED;
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/warehouse/Warehouse.java:275]: just a remind, we also need to persist this state and release all the cluster resource here

File: fe/fe-core/src/main/java/com/starrocks/warehouse/Warehouse.java [hunk 17]
```diff
@@ -203,15 +203,18 @@ public void suspendSelf(boolean isReplay) {
         if (!isReplay) {
             releaseComputeNodes();
         }
+        clusters.clear();
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/warehouse/Warehouse.java:206]: should use try..finally for lock/unlock

File: fe/fe-core/src/main/java/com/starrocks/warehouse/Warehouse.java [hunk 18]
```diff
@@ -203,15 +203,18 @@ public void suspendSelf(boolean isReplay) {
...
 
     public void resumeSelf(boolean isReplay) {
         writeLock();
+        Map<Long, Cluster> clusterMap = new HashMap<>();
         if (!isReplay) {
-            applyComputeNodes();
+            clusterMap = applyComputeNodes();
         }
         this.state = WarehouseState.RUNNING;
+        clusters = clusterMap;
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/warehouse/Warehouse.java:217]: should use try..finally for lock/unlock

File: fe/fe-core/src/main/java/com/starrocks/task/AgentBatchTask.java [hunk 19]
```diff
@@ -91,17 +93,13 @@ public void addTask(AgentTask agentTask) {
             return;
         }
         long backendId = agentTask.getBackendId();
+        TNetworkAddress addr = agentTask.getBackendAddr();
         List<AgentTask> tasks = backendIdToTasks.computeIfAbsent(backendId, k -> new ArrayList<>());
         tasks.add(agentTask);
+        List<AgentTask> tasks2 = backendAddrToTasks.computeIfAbsent(addr, k -> new ArrayList<>());
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/task/AgentBatchTask.java:99]: what is tasks2?

File: fe/fe-core/src/main/java/com/starrocks/lake/StarOSAgent.java [hunk 20]
```diff
@@ -395,6 +396,31 @@ public long getPrimaryBackendIdByShard(long shardId) throws UserException {
...
+        // for debug
+        LOG.info("replicas size is {}", replicas);
+
+        try (LockCloseable lock = new LockCloseable(rwLock.writeLock())) {
+            for (ReplicaInfo replicaInfo : replicas) {
+                if (replicaInfo.getReplicaRole() == ReplicaRole.PRIMARY) {
+                    WorkerInfo workerInfo = replicaInfo.getWorkerInfo();
+                    String workerAddr = workerInfo.getIpPort();
+                    String host = workerAddr.split(":")[0];
+                    int bePort = Integer.parseInt(workerInfo.getWorkerPropertiesMap().get("BePort"));
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/lake/StarOSAgent.java:414]: BePort -> be_port

**Pull request closed without merging**